### PR TITLE
Make it work under windows.

### PR DESCRIPTION
**Pull request by inter1965**
Add some improvements including options.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (alhassy) [inline .vscode/settings.json]: What's this file for?

I'm not too familiar with all the config options.

File: E-snippets.json [hunk 2]
```diff
@@ -208,12 +208,12 @@
   " Get the name of the current file, editor, as a string.": {
     "prefix": "E.currentFileName ",
     "body": "E.currentFileName(${1:})$0",
-    "description": " Get the name of the current file, editor, as a string.\n *\n * For example, within my `~/init.js` pressing `cmd+e` on the following:\n * ```\n * E.currentFileName() //  ⇒  /Users/musa/init.js\n * ```\n "
```
Comment (alhassy) [inline E-snippets.json]: This just has `init.js` to `.init.js` renaming.

Okie dokies.

File: extension.js [hunk 3]
```diff
@@ -103,34 +105,36 @@ E.internal.eval.require = (pkg, explicitPath = 'index.js') => {
 
 commands["Open the tutorial; I'd like to learn more about using cmd+E!"] = E => {
   E.shell(
-    'rm ~/Downloads/tutorial.js; curl -o ~/Downloads/tutorial.js https://raw.githubusercontent.com/alhassy/easy-extensibility/main/tutorial.js'
+    `rm ${E.expanduser("~/Downloads/tutorial.js")}; curl -o ${E.expanduser("~/Downloads/tutorial.js")} https://raw.githubusercontent.com/alhassy/easy-extensibility/main/tutorial.js`
```
Comment (alhassy) [inline extension.js]: Great idea!

File: extension.js [hunk 4]
```diff
@@ -35,6 +35,8 @@ require('fs').writeFileSync('/Users/musa/easy-extensibility/E-snippets.json', JS
 // vscode, E, commands ================================================================================
 
 const vscode = require('vscode')
+const conf = vscode.workspace.getConfiguration("easy-extensibility").get
+const rcfile = conf("rcFile")
```
Comment (alhassy) [inline extension.js]: It looks like you're using `rcFile` as the config setting to avoid the hardcoded name `init.js`.

1. Why the name `rcFile`? In particular, what does  the prefix`rc` stand for?
2. Would you please tell me a bit more about line 38. Does this get our/easy-extensibility's internal configuration or the users `settings.json`?

File: extension.js [hunk 5]
```diff
@@ -103,34 +105,36 @@ E.internal.eval.require = (pkg, explicitPath = 'index.js') => {
...
+    // That is, the user's `~/.init.js` file can make liberal use of `await` clauses as a syntactic sugar for an ambient async IIFE.
     eval(`(async () => { ${text} })()`)
-    E.message('~/init.js loaded!')
+    E.message(`${rc} loaded!`)
   })
 }
 
 //!! Actually load the file upon startup!
-commands['Reload ~/init.js file'](E)
+commands['Reload rc file'](E)
```
Comment (alhassy) [inline extension.js]: 👀 "rc" ? What does this stand for?

File: extension.js [hunk 6]
```diff
@@ -166,6 +170,11 @@ function activate(context) {
     )
   register('evaluateSelection')
   register('executeRegisteredCommand')
+
+  let initialization = () => {
+    E.node_path = E.shell(conf("npmCommand") + " root -g")
```
Comment (alhassy) [inline extension.js]: Nice!

File: extension.js [hunk 7]
```diff
@@ -166,6 +170,11 @@ function activate(context) {
...
+  }
+  initialization()
```
Comment (alhassy) [inline extension.js]: Why make the initialization code into a top-level function, only to invoke it immediately?

Why not just invoke it immediately? I.e., inline line 177 with the definition on line 175?

Or do you anticipate it can be used as a hooking mechanism? If so, could you elaborate on your vision here 😁

File: init.js [hunk 8]
```diff
@@ -11,20 +11,20 @@
...
 
 // It is encouraged to keep this file under version control, then make a symbolic link;
-// e.g.,   ln  -s  ~/my-cool-repo/init.js  ~/init.js
+// e.g.,   ln  -s  ~/my-cool-repo/init.js  ~/.init.js
 
 // ==================================================================================
 // ========= 🚀 Whenever we open VSCode, let's see a motivating message! 💪 ==========
 // ==================================================================================
-let welcome = `Welcome ${process.env.USER}! Today is ${E.shell('date')}!`
+let welcome = `Welcome ${E.username()} on ${process.platform}! Today is ${E.date()}!`
```
Comment (alhassy) [inline init.js]: I'm not sure about this change.

----

Let me defend the original, and then you defend the alteration and we'll take it from there 😁

----

The original is short & sweet:
- It involves only 2 interpolations.
- The first demonstrates the easy access to `process` ---which would be nice for existing NodeJS users to see.
- The second demonstrates the easy access to running shell commands ---which is a powerful feature for people to see right away.

While the alteration does a good job of lifting there to functions, for increased readability; I worry that the abstractions belie/obscure the power of the `E` API itself.

File: init.js [hunk 9]
```diff
@@ -520,3 +520,17 @@ commands['Define word'] = {
 // this essentially removes a lot of the distractions from your view in VS Code.
 // (As always, toggle this with Cmd+Shift+P ---or Ctrl+X+Z)
 E.executeCommand('workbench.action.toggleZenMode')
+
+function quoteSelection(quote = "\"", sep = " ") {
```
Comment (alhassy) [inline init.js]: Some discussion on the merits of including these would be appreciated.
- Even a single-line comment justifying why you want these would be nice.
- Also, you can replace your 5-line function with a single line using `E.replaceSelection`.
   + A nearly identical implementation to your function already exists here, https://github.com/alhassy/easy-extensibility/blob/main/init.js#L42, so maybe you could consider altering that 🤷‍♂️

File: package.json [hunk 10]
```diff
@@ -1,90 +1,126 @@
 {
-  "name": "easy-extensibility",
```
Comment (alhassy) [inline package.json]: Did anything actually change here, or is it just linting/formatting?
- If it's the latter, then please revert the change ---i.e., don't commit this. Thanks.

File: vscodejs/index.js [hunk 11]
```diff
@@ -20,10 +20,25 @@
  * ```
  * This pattern is captured by `E.withEditor`.
  */
+const path = require("path")
```
Comment (alhassy) [inline vscodejs/index.js]: I think all of these are great additions! 🚀

Please expose them at the top-level:

```js
/** This accomplishes goals ABC. Example uses are as follows... */
E.getConfigDir = () => {...}

/** This accomplishes goals ABC. Example uses are as follows... */
E.isWindows = () => ...

/** This accomplishes goals ABC. Example uses are as follows... */
E.userConfigDir = () => ...

/** This accomplishes goals ABC. Example uses are as follows... */
E.userDirc = () => ...
```

File: vscodejs/index.js [hunk 12]
```diff
@@ -513,7 +536,7 @@ module.exports = vscode => {
     }
     if (options.name) {
       E.shell(`touch ${options.name}`)
-      await vscode.window.showTextDocument(vscode.Uri.file(options.name.replace(/~/g, process.env.HOME)))
+      await vscode.window.showTextDocument(vscode.Uri.file(E.expanduser(options.name)))
```
Comment (alhassy) [inline vscodejs/index.js]: Good catch!

File: vscodejs/index.js [hunk 13]
```diff
@@ -805,13 +828,24 @@ module.exports = vscode => {
...
    * ```
    */
   E.currentDirectory = () => E.currentFileName().split('/').slice(0, -1).join('/')
 
+  /** Expand user directory
+   *  expand "~" to actual path.
+   */
+  E.expanduser = ( pathstr ) => {
+    if( pathstr.indexOf("~") === 0) {
+      return path.normalize(path.join(userDir, pathstr.slice(2)))
```
Comment (alhassy) [inline vscodejs/index.js]: Why is this better than plain old `pathstr.replace(/~/g, process.env.HOME)`?

Could you please adorn the documentation of `expandUser` with some examples.

**Review by alhassy — CHANGES_REQUESTED** [review]
Thanks for the PR!

Lots of good ideas here; I've left some questions around. Once we've discussed those, then I'm happy to merge 🥳

File: .vscode/settings.json (hunk 1 above)
Comment (inter1965) [reply .vscode/settings.json]: Sorry, that's my miss.
It should go into .gitignore.

File: E-snippets.json (hunk 2 above)
Comment (inter1965) [reply E-snippets.json]: Just make it portable.

File: extension.js (hunk 4 above)
Comment (inter1965) [reply extension.js]: Please refer to [https://stackoverflow.com/questions/11030552/what-does-rc-mean-in-dot-files](url)
The ‘rc’ suffix goes back to Unix's grandparent, CTSS. It had a command-script feature called "runcom". Early Unixes used ‘rc’ for the name of the operating system's boot script, as a tribute to CTSS runcom.

File: extension.js (hunk 5 above)
Comment (inter1965) [reply extension.js]: The ‘rc’ suffix goes back to Unix's grandparent, CTSS. It had a command-script feature called "runcom". Early Unixes used ‘rc’ for the name of the operating system's boot script, as a tribute to CTSS runcom.

File: extension.js (hunk 7 above)
Comment (inter1965) [reply extension.js]: Since under windows it's relatively expensive to start a new process, hence here's i just cache the result of the npm module path for faster execution.

File: init.js (hunk 8 above)
Comment (inter1965) [reply init.js]: Because the environmental variables are different under individual platform, .e.g. it's USER under unix like system while it's USERNAME under windows.
I think if you want to exemplify the power of this extension that we have to make sure all the code are runnable under different os, otherwise it would only frustrate people.
for example, the date command is `date +%H:%M:%S`  under unix like os while it's `data /T` under windows.

File: init.js (hunk 9 above)
Comment (inter1965) [reply init.js]: Thanks, this is just making more customization to the quote operation.
Given a extra function here it is just to keep the code concise since it is going to be reused by the following three E commands.
Use along with `E.replaceSelection` will make it have to extract the selection again since we are doing some extra transformation here; hence here I have to expand it like the current code; otherwise we have to refactor the `E.replaceSelection` function.
For example, if one just want to turn the command line "-i a.file -o b.out -c xxx -d yyy" into some json array ["-i", "a.file", "-o", "b.out", "-c", "xxx", "-d", "yyy"] which existed in the tasks.json

File: package.json (hunk 10 above)
Comment (inter1965) [reply package.json]: This should disappear once the .vscode folder is included in the .gitignore file.

File: vscodejs/index.js (hunk 11 above)
Comment (inter1965) [reply vscodejs/index.js]: Got it.

File: vscodejs/index.js (hunk 13 above)
Comment (inter1965) [reply vscodejs/index.js]: Please consider some extreme case like a path "~/foo/~bar/".
expandUser is a function mimic the python's os.path.expanuser

> os.path.expanduser(path)
> On Unix and Windows, return the argument with an initial component of ~ or ~user replaced by that user’s home directory.
> 
> On Unix, an initial ~ is replaced by the environment variable HOME if it is set; otherwise the current user’s home directory is looked up in the password directory through the built-in module [pwd](https://docs.python.org/3/library/pwd.html#module-pwd). An initial ~user is looked up directly in the password directory.
> 
> On Windows, USERPROFILE will be used if set, otherwise a combination of HOMEPATH and HOMEDRIVE will be used. An initial ~user is handled by checking that the last directory component of the current user’s home directory matches USERNAME, and replacing it if so.
> 
> If the expansion fails or if the path does not begin with a tilde, the path is returned unchanged.
>
